src/flist/db.rs

```diff
@@ -0,0 +1,120 @@
+use anyhow::{bail, Result};
+use bytes::Buf;
+use std::io;
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+pub struct MetadataDbMgr {
+    // root directory where all
+    // the working file of the module will be located
+
+    // underneath are the path for each
+    // sub folder used by the flist module
+    flist: PathBuf,
+}
+
+impl MetadataDbMgr {
+    pub async fn new<P: AsRef<Path>>(flist: P) -> Result<Self> {
+        Ok(Self {
+            flist: flist.as_ref().to_path_buf(),
+        })
+    }
+    pub async fn get<T: AsRef<str>>(&self, url: T) -> Result<PathBuf> {
+        let url = url.as_ref();
+        let hash = self.hash_of_flist(url).await?;
+        let path = self.flist.join(&hash.trim());
+        match fs::File::open(&path).await {
+            Ok(_) => {
+                //Flist already exists let's check it's md5
+                if self.compare_md5(hash, &path) {
+                    return Ok(path);
+                } else {
+                    self.download_flist(url).await
+                }
+            }
+            Err(error) => match error.kind() {
+                io::ErrorKind::NotFound => self.download_flist(url).await,
+
+                _ => bail!(
+                    "error reading flist file: {}, error {}",
+                    &path.display(),
+                    error
+                ),
+            },
+        }
+    }
+    pub fn compare_md5<T: AsRef<str>, P: AsRef<Path>>(&self, hash: T, path: P) -> bool {
+        // create a Md5 hasher instance
+        let calculated_hash =
+            checksums::hash_file(path.as_ref(), checksums::Algorithm::MD5).to_lowercase();
+        calculated_hash == hash.as_ref()
+    }
+    // downloadFlist downloads an flits from a URL
+    // if the flist location also provide and md5 hash of the flist
+    // this function will use it to avoid downloading an flist that is
+    // already present locally
+    pub async fn download_flist<T: AsRef<str>>(&self, url: T) -> Result<PathBuf> {
+        let url = url.as_ref();
+        // Flist not found or hash is not correct, let's download
+        let mut resp = reqwest::get(url).await?.bytes().await?.reader();
+        let mut builder = tempfile::Builder::new();
+        let mut file = builder.suffix("_flist_temp").tempfile_in(&self.flist)?;
```

Comment:
no need for the `builder` variable. 
```suggestion
let mut file = tempfile::Builder::new()
            .suffix("_flist_temp")
            .tempfile_in(&self.flist)?;
```